### PR TITLE
API: added OpenAPI document

### PR DESCRIPTION
**Pull request by baz2142**

### CODE REVIEW

**Review by juliuskoskela — COMMENTED** [review]
I suggest we don't commit the generated documentation in the repository itself. Usually generated documentation is hosted separately. Reasonging:

- Repository size and clutter.
- Git diff clutter.
- Unnecessary merge conflicts and other chaos.

**Comment by juliuskoskela** [comment]
When trying to generate from the openapi.yaml using [openapi-generator-cli](https://github.com/OpenAPITools/openapi-generator) I get the following error:

```
~/Repos/vmd-baz/vmd-api/api main*
❯ openapi-generator-cli generate -g rust -i openapi.yaml -o test
Exception in thread "main" org.openapitools.codegen.SpecValidationException: There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 1, Warning count: 0
Errors:
	-attribute info.license.identifier is missing

	at org.openapitools.codegen.config.CodegenConfigurator.toContext(CodegenConfigurator.java:620)
	at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput(CodegenConfigurator.java:647)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:479)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
```

**Comment by juliuskoskela** [comment]
Need to update license like so:

```yaml
  license:
    name: Apache License 2.0
    url: https://www.apache.org/licenses/LICENSE-2.0.html
    identifier: Apache-2.0
```

File: vmd-api/api/openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+  url: https://localhost/api/v1
+- description: The test server
+  url: https://localhost/api/test/v1
+tags:
+- description: Operations of Virtual Machines
+  name: vm
+paths:
+  /vm/list:
+    get:
+      description: Return a IDs list of all Virtual Machines
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Return a IDs list of all Virtual Machines/Gets list of IDs for all virtual machines/

File: vmd-api/api/openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,336 @@
+openapi: 3.1.0
+info:
+  contact:
+    email: nikita.bazulin@unikie.com
+  description: This is Control Panel API based on the OpenAPI 3.1 specification of
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/This is Control Panel API based on the OpenAPI 3.1 specification of../OpenAPI specification for virtual machine management on Ghaf platform/

File: vmd-api/api/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    the GHAF project
+  license:
+    name: Apache License 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  title: Virtual Machine Manager API of the GHAF project - OpenAPI 3.1
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Manager/Management/

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: s/of the GHAF project/on the Ghaf platform/

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: Please note that Ghaf is not an acronym for anything so not all caps - it's an arab word for a specific tree.

File: vmd-api/api/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+- description: The test server
+  url: https://localhost/api/test/v1
+tags:
+- description: Operations of Virtual Machines
+  name: vm
+paths:
+  /vm/list:
+    get:
+      description: Return a IDs list of all Virtual Machines
+      operationId: getVmList
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:26]: Do we need `operationId` (optional field) for anything else than documentation generation? Is it used for code generation (e.g. function names get generated based on it) on client or server side? @juliuskoskela 
If the latter is yes, we should check the case for code styling. RFC3986 recommends hyphen as separator (spinal-case instead of camel-case).

Comment (vilvo) [reply vmd-api/api/openapi.yaml:26]: Above comment is related to all of the following `operationId`s

File: vmd-api/api/openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                items: {}
+          description: IDs list of all Virtual Machines
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/IDs list of all Virtual Machines/list of IDs for all virtual machines/

File: vmd-api/api/openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        default:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Get VMs list
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Get VMs list/Get list IDs for all virtual machines/

File: vmd-api/api/openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+      tags:
+      - vm
+  /vm/info/{id}:
+    get:
+      description: Returns information about the Virtual Machine by ID
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Returns information about the Virtual Machine by ID/Gets the virtual machine information by ID/

File: vmd-api/api/openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+      operationId: getVmInfoById
+      parameters:
+      - description: An ID of the Virtual Machine that needs to be fetched
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/An ID of ../The ID of the virtual machine/ # please drop "that needs to be fetched"

File: vmd-api/api/openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+        style: simple
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/VirtualMachineInfo'
+          description: successful operation
+        "400":
+          description: Invalid Virtual Machine ID supplied
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: please drop "supplied"

File: vmd-api/api/openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+        "404":
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:64]: What is actually the semantic difference between 400 (invalid) and 404 (not found)?

File: vmd-api/api/openapi.yaml [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+          description: Virtual Machine ID not found
+        default:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Finds Virtual Machine info by ID
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Finds ../Gets virtual machine info by ID/

File: vmd-api/api/openapi.yaml [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+      tags:
+      - vm
+  /vm/{action}/{id}:
+    put:
+      description: Make an action to a Virtual Machine by it's ID
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Make an ../Control the virtual machine by ID/

File: vmd-api/api/openapi.yaml [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+      operationId: vmAction
+      parameters:
+      - description: "Action, which will be applied to a Virtual Machine"
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Action .../Control action to the virtual machine/

File: vmd-api/api/openapi.yaml [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+        in: path
+        name: action
+        required: true
+        schema:
+          enum:
+          - start
+          - stop
+          - pause
+        style: simple
+      - description: Virtual Machine's ID
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Virtual Machine's ID/ID of the virtual machine/

File: vmd-api/api/openapi.yaml [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+        example: VM#101
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: The `VM#101` ID encoding is redundant as the `VM` is clear from the URI context, also having it with the `VM` prefix would require that the server would return the list of IDs with all prefixed (also redundant). Let's drop the `VM`-prefix and only use the IDs. As with numeric unique IDs and we don't yet know how we generate the unique VM IDs on the server side so let's leave this example out for now.

File: vmd-api/api/openapi.yaml [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+        style: simple
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/VirtualMachineInfo'
+          description: successful operation
+        "400":
+          description: Invalid Virtual Machine ID supplied
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: Answering my own earlier question - let's define "Invalid" ID as containing:
- any non-numeric characters
- too small (smaller than to be defined)
- too big (larger than to be defined)
@juliuskoskela

File: vmd-api/api/openapi.yaml [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+        "404":
+          description: Virtual Machine ID not found
+        default:
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:111]: Let's define as "Valid but not found"

File: vmd-api/api/openapi.yaml (hunk 4 above)
Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml:26]: The generator seems to be smart about this. On the rust side getVmLists generates:

```rust
pub async fn get_vm_list(configuration: &configuration::Configuration, ) -> Result<serde_json::Value, Error<GetVmListError>>;
```

File: vmd-api/api/openapi.yaml [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Make an action to a Virtual Machine by it's ID
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: s/Make an .../Request control action to the virtual machine by the ID"

File: vmd-api/api/openapi.yaml [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+        required: true
+        schema: {}
+        style: simple
+      responses:
+        "200":
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/VirtualMachineInfo'
+          description: successful operation
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:106]: Will the ID of the virtual machine change with `stop`/`start`?
We must define that as if the ID is based on outside VM PID, we have to maintain IDs on VMD which will make VMD stateful, not stateless. I would propose we use the PID (process ID) of the VM as it is not leaking any critical information outside host and will not require VMD server to maintain any other list of VM IDs.

File: vmd-api/api/openapi.yaml [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+          networkIo: ""
+        os:
+          name: Ubuntu
+          version: 23.04
+        devices:
+          list: ""
+        hypervisorDetails:
+          emulator: /usr/bin/qemu-system-x86_64
+          hypervisor: KVM
+          arch: armv8
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: Let's use only relevant examples on arch. `aarch64`etc.

File: vmd-api/api/openapi.yaml [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+        cpu:
+          cores: 4
+          threads: 8
+          model: NVidia Tegra X1
+          sockets: 1
+          logicalCount: 8
+        basicDetails:
+          name: GUI VM
+          description: GUI Virtual Machine
+          id: VM#101
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: See my comments on IDs

File: vmd-api/api/openapi.yaml [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+          state: Running
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:165]: Valid states must be defined.

File: vmd-api/api/openapi.yaml [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+        startTime: ""
+        runningTime: ""
+      properties:
+        basicDetails:
+          $ref: '#/components/schemas/VirutalMachineBasicDetails'
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: Please fix typos: s/Virutal/Virtual/

File: vmd-api/api/openapi.yaml [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+        state: Running
+      properties:
+        id:
+          example: VM#101
+        state:
+          enum:
+          - Stopped
+          - Starting
+          - Running
+          - Stopping
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:204]: Virtual machines can be in error states as well.

File: vmd-api/api/openapi.yaml [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+          $ref: '#/components/schemas/VirutalMachinePerformance'
+        devices:
+          $ref: '#/components/schemas/VirutalMachineDevices'
+      required:
+      - basicDetails
+      - cpu
+      - memory
+      - os
+    VirutalMachineBasicDetails:
+      example:
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml:191]: Can we cross-reference to earlier part of the specification? Instead of copy-paste (with the same issues).

File: vmd-api/api/openapi.yaml [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Make an action to a Virtual Machine by it's ID
+      tags:
+      - vm
+components:
+  schemas:
+    Rational:
```
Comment (vilvo) [inline vmd-api/api/openapi.yaml]: The rationale of this `Rational` is unclear to me. You want to return numbers that are represented as quotients of integers? Is this absolutely required and what's the use case where we need this level of accuracy on the client side? Option is to use integers and round the to floor or ceiling (defined) on the server side.

**Review by vilvo — CHANGES_REQUESTED** [review]
Please address review comments.
Also, let's not included generated documentation but let's cover auto-generation (or instructions for manual generation) in separate issue.

**Comment by vilvo** [comment]
Nice work with the first version for comments @baz2142 !

**Comment by juliuskoskela** [comment]
Realized a repeating typo, check the files and filenames for `virutal` instead of `virtual`.

**Comment by vilvo** [comment]
@baz2142 - also please add DCO, Developer Certificate of Origin, to your commits - https://wiki.linuxfoundation.org/dco#developer_certificate_of_origin_dco

File: vmd-api/api/openapi.yaml (hunk 1 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Why use Gets, but not Returns?

I was following the style in the reference example:
https://github.com/swagger-api/swagger-petstore/blob/master/src/main/resources/openapi.yaml#L325

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: My 2 cents:
- shorter (preferred)
- aligned with the verb (get)

File: vmd-api/api/openapi.yaml (hunk 26 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: I was thinking about memory usage. E.g. "100MB/512MB". Actually, now I think the denominator will be easiar deduced from VirutalMachineMemory.currentAllocation

File: vmd-api/api/openapi.yaml (hunk 15 above)
Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml]: Should we use https://docs.rs/uuid/latest/uuid/?

File: vmd-api/api/openapi.yaml (hunk 24 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:204]: Tell me more about this: will it be VM is the Running state, which encountered problem? Or it will be turned off to the Stopped state?

The main point of my question: is it mutual exclusion, having an error and any of the states? If VM with an error may be in Running or Stopped states, maybe we should add some field about errors alongside with those states?

File: vmd-api/api/openapi.yaml (hunk 19 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:106]: I wasn't thinking ID of VM will be changed. KVM has UUID, which is not affected by start/stop. Am I right, @juliuskoskela?

I was thinking it (ID) will be an UUID from KVM. But I'm not aware of implementation details of the server side

File: vmd-api/api/openapi.yaml (hunk 22 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:165]: Explain it, please. It is a valid string state from enum VirtualMachineBasicDetails.state.enum

File: vmd-api/api/openapi.yaml (hunk 24 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:204]: I'm thinking Linux process states - there's also sleep and zombie (defunct) states. In essence, a VM run with VMM is a process in Linux.

File: vmd-api/api/openapi.yaml (hunk 15 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml]: `uuid` (any) adds a mapping layer between `uuid` and `pid` (process id). The question is if we want that and what are the benefits the `uuid` brings (other than hide `pid` from client)? It will also require definitions like will `uuid` for a VM be persistent over restart (stop-start) cycles etc. 
With `pid` only, Linux will manage the IDs.

Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml]: But PIDs wouldn't be persistent either, right?

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: No. And implementing superficial persistence ids in server goes against representational state transfer statelessness design https://en.wikipedia.org/wiki/Representational_state_transfer#Statelessness

File: vmd-api/api/openapi.yaml (hunk 16 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: We can set the limits inside this document later

File: vmd-api/api/openapi.yaml (hunk 15 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: I was thinking KVM gives us uuid of each VM, isn't it?
In my head it was the ID, which will be used

File: vmd-api/api/openapi.yaml (hunk 17 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:111]: Do you mean error 404?
I was thinking it will be good to have "default error", which will contains details inside ErrorModel

File: vmd-api/api/openapi.yaml (hunk 24 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:204]: I think we need to think about this. The process itself and the VM can be in "normal" state, but the process can write something to stderr. So will be it error state or not?

Maybe we should split states: VM state, process state, program state

File: vmd-api/api/openapi.yaml (hunk 23 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 1 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 2 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 3 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 5 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 6 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 7 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 8 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 10 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml:64]: I guess you have answered below

File: vmd-api/api/openapi.yaml (hunk 11 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 12 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 13 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 14 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 18 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 20 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 21 above)
Comment (baz2142) [reply vmd-api/api/openapi.yaml]: Done

File: vmd-api/api/openapi.yaml (hunk 24 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:204]: Hmm, you might be mixing process (actual operating system concept) with meta concept of virtual machine (VM, also just a process and only a process from the host operating system perspective). The VM can be considered "a special process" as it uses resources via KVM (other, normal host OS processes do not use hardware resources via KVM but only via host OS kernel). Still, all processes are just processes and the VM is not different.

In practice:

```
$ ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0 172752 11264 ?        Ss   09:28   0:01 /run/current-system/systemd/lib/systemd/systemd
... <rest of the processes>

$ man ps
PS(1)                                                                                             User Commands                                                                                             PS(1)

NAME
       ps - report a snapshot of the current processes.
...
PROCESS STATE CODES
       Here are the different values that the s, stat and state output specifiers (header "STAT" or "S") will display to describe the state of a process:

               D    uninterruptible sleep (usually IO)
               I    Idle kernel thread
               R    running or runnable (on run queue)
               S    interruptible sleep (waiting for an event to complete)
               T    stopped by job control signal
               t    stopped by debugger during the tracing
               W    paging (not valid since the 2.6.xx kernel)
               X    dead (should never be seen)
               Z    defunct ("zombie") process, terminated but not reaped by its parent
```

So, I am arguing that we should stick to host OS process states for VMs. In particular, the VM states reported to the client are only the subset of all host OS processes. Namely, processes run with VMMs (like `qemu`, `crosvm`, and other VMMs). 
There's no way of the vmd server (or host OS) knowing what actually is the state of the VMs beyond `PROCESS STATE CODES` without some intelligent debugging of the VM process (we won't do that).

**Comment by vilvo** [comment]
@baz2142 - please check commit signing for "Verified" commits. More info at https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

File: vmd-api/api/openapi.yaml (hunk 9 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml]: this was also Done in [4a67b9d](https://github.com/tiiuae/vmd/commit/4a67b9def64aacc30f13dee1391508df8074e362)

File: vmd-api/api/openapi.yaml (hunk 22 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:165]: My proposal. https://github.com/tiiuae/vmd/pull/24#discussion_r1144643441

File: vmd-api/api/openapi.yaml (hunk 19 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:106]: What exactly is the proposed "UUID from KVM"? Can you please link to relevant part of Linux kernel documentation?

File: vmd-api/api/openapi.yaml (hunk 16 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml]: https://github.com/tiiuae/vmd/issues/30

File: vmd-api/api/openapi.yaml (hunk 15 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml]: KVM is a kernel virtualization module and I could not find `uuid`s that KVM would give to userland. It's basically so that a VMM (process) runs virtual machines using `ioctl`s to KVM API. What is seen on the userland side is process ID of the VMM process that is running the virtual machine.
If there's such uuid somewhere, I'd be happy to be proven wrong but please refer to documentation for exactly where this comes from.

File: vmd-api/api/openapi.yaml (hunk 10 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:64]: I had to guess, as well :-) 
I will be very explicit here. Let's think this OpenAPI specification as such that we give the reader (also future readers) these answers and definitions. Based on your comment, I'm guessing you agree with my guesses :-) So, would you please add the definitions of 404 (Invalid) and 404 (Not found) to the descriptions? Thank you.

File: vmd-api/api/openapi.yaml (hunk 4 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:26]: @baz2142 this question was not yet addressed: "Do we need operationId (optional field) for anything else than documentation generation?"

File: vmd-api/api/openapi.yaml (hunk 17 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:111]: I mean that let's add in the description following definition (if you agree):
Valid ID is VM ID that is within valid boundaries (defined later by https://github.com/tiiuae/vmd/issues/30) but the ID is not found by the VMD. "Not found" could be a reason where the client tries to get unauthorized access to IDs that are not IDs of available VMs.

File: vmd-api/api/openapi.yaml (hunk 15 above)
Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml]: From KVM's perspective I think the we have name and id (which is a u32).

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: When creating a VM via KVM API on Linux, we have fd, file descriptor.
screencapture
The VMM process (PID) can have a set of file descriptors under `/proc/<PID>/fd/` as the VM may access other resources, via their fds, as well.
Referring to VM fd (as u32) makes little sense as it's meant as the VMM (e.g. `crosvm`) internal. See more info on how KVM uses fds - https://www.kernel.org/doc/html/latest/virt/kvm/api.html#file-descriptors

Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml]: I wasn't referring to the fd, but the PID which is a u32.

But kvm documentation states that a vm is only associated with the fd and not with the process (PID).

_It is important to note that although VM ioctls may only be issued from the process that created the VM, a VM’s lifecycle is associated with its file descriptor, not its creator (process). In other words, the VM and its resources, including the associated address space, are not freed until the last reference to the VM’s file descriptor has been released. For example, if fork() is issued after ioctl(KVM_CREATE_VM), the VM will not be freed until both the parent (original) process and its child have put their references to the VM’s file descriptor._

[If we run a program like `virsh` we get fields ID and Name.](https://www.cyberciti.biz/faq/linux-list-a-kvm-vm-guest-using-virsh-command/)

Reading further into it [RedHat documentation talks about a Domain ID](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_deployment_and_administration_guide/sect-managing_guest_virtual_machines_with_virsh-retrieving_guest_virtual_machine_information) but it's unclear if this is a mapping done by the `virsh` program is this ID somehow associated with the file descriptor.

Furthermore KVM states about forking or dupping child processes:

_Because a VM’s resources are not freed until the last reference to its file descriptor is released, creating additional references to a VM via fork(), dup(), etc… without careful consideration is strongly discouraged and may have unwanted side effects, e.g. memory allocated by and on behalf of the VM’s process may not be freed/unaccounted when the VM is shut down._

So what I gather is that we definitely don't use information of the process such as the PID for identification, but using the file descriptor is possibly problematic if we spawn child processes.

The `virsh` program has the following identifiers for a vm:

- Domain name
- Domain ID
- UUID

Example:

```
# virsh dominfo guest1
Id:             8
Name:           guest1
UUID:           90e0d63e-d5c1-4735-91f6-20a32ca22c48
OS Type:        hvm
State:          running
CPU(s):         1
CPU time:       32.6s
Max memory:     1048576 KiB
Used memory:    1048576 KiB
Persistent:     yes
Autostart:      disable
Managed save:   no
Security model: selinux
Security DOI:   0
Security label: system_u:system_r:svirt_t:s0:c552,c818 (enforcing)
```

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: Ok. I've been arguing for PID as it's independent of VMM (QEMU, rust-vmm-based VMMs, ...). However, as you introduced the virsh example, there are alternatives to `vmd` on managing guest VMs, often designed for cloud, rather than edge, virtualization management.

Please note that ghaf uses microvm.nix generated systemd units. This is important distinction to virsh (and others).

See:
https://github.com/tiiuae/ghaf/blob/main/microvmConfigurations/netvm/default.nix
https://astro.github.io/microvm.nix/microvm-command.html#enabling-microvm-autostart
https://astro.github.io/microvm.nix/host-systemd.html

Target VMM is rust-vmm-based `crosvm`. In some cases we can use QEMU or other VMMs to develop and test concepts.

In practice, this means what I tried to describe earlier:
1. fd is `crosvm` internal. We can't use the fd to issue ioctls outside `crosvm` ("strongly discouraged")
2. We can analyze the `crosvm` process further based on `PID` - e.g. via `procfs`. This is valid, allowed and has no issues nor side effects.
3. The only case where we would use KVM API ioctls is to resolve the virtualization capabilities of the underlying HW when the `vmd` is started. This proved to be poorly supported by different HW so we may abandon direct dependency to KVM API and introduce vmd server internal dependency that better suits this purpose. Another key reason is that "no network on host requirement" allows vmd server to do only indirect queries to host (e.g. via shared memory).
So, we are not planning to use KVM API assigned `fd` of VM directly - that will be owned by the VMM on host.

In other words - analyzing running VMs using tools alternative to microvm.nix+VMM (like virsh) via fd is risky even if it would work (as you quoted the KVM documentation).

Given that, virsh concepts like "Domain name", "Domain ID" and "UUID" are irrelevant within this design context _unless_ you propose explicitly:
1. why those concepts are must in our context
2. what is the exact mechanism to generate those given our dependencies and constraints described above

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: Additionally, `PID` does not depend on `crosvm` nor other VMMs. We can get key information via `PID` for all VMM processes and then some (if needed).

Comment (vilvo) [reply vmd-api/api/openapi.yaml]: Study the usage of `pidfd` to implement VM ID via API

File: vmd-api/api/openapi.yaml (hunk 4 above)
Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml:26]: My understanding is that it's what the function names in the generated API are derived from?

File: vmd-api/api/openapi.yaml (hunk 24 above)
Comment (juliuskoskela) [reply vmd-api/api/openapi.yaml:204]: Would this be a direct mapping of the the PROCESS_STATE_CODES or are some of the states redundant or should be grouped together in our context?

File: vmd-api/api/openapi.yaml (hunk 19 above)
Comment (vilvo) [reply vmd-api/api/openapi.yaml:106]: Resolving proposed UUID from KVM as invalid.
